bastionzero/zli
# Collapse `zli policy` related commands

## Description of the change

There are currently 6 policy-related commands within the zli. These commands are as follows:
`zli policy [type]`
`zli describe-cluster-policy <clusterName>`
`zli user [policyName] [idpEmail]`
`zli group [policyName] [groupName]`
`zli targetuser <policyName> [user]`
`zli targetgroup <policyName> [group]`

These commands are now collapsed into the main `zli policy` command as subcommands. They now look like so:
`zli policy help`
      `zli policy list [type]`
      `zli policy describe-cluster-policy <clusterName>`
      `zli policy users`
      `zli policy add-user <policyName> <idpEmail>`
      `zli policy delete-user <policyName> <idpEmail>`
      `zli policy groups`
      `zli policy add-group <policyName> <groupName>`
      `zli policy delete-group <policyName> <groupName>`
      `zli policy targetusers <policyName>`
      `zli policy add-targetuser <policyName> <targetUser>`
      `zli policy delete-targetuser <policyName> <targetUser>`
      `zli policy targetgroups <policyName>`
      `zli policy add-targetgroup <policyName> <targetGroup>`
      `zli policy delete-targetgroup <policyName> <targetGroup>`

Main goals:
1. Refactor the zli policy commands. 
2. Abstract the fetch data middleware from the cli-driver. Fetching the data needed doesn't need to happen for every command. So now, the data is fetched in the handlers that need the data and **only** the data that's needed is fetched. 


## Testing

Manual test cases:
I've created a `Kubernetes` policy in `cloud-staging` called `sai-policy-refactoring-test`. The reason for this policy specifically is to be able to test target groups as well with the same policy. 

1. zli policy help
You can also run the help command with any of the subcommands
Expected: list all the new policy subcommands that are listed above

3. zli policy list
You can also run this using the --type flag for specific policy types
Expected: list all policies in default or the policies filtered by policy type

4. zli policy describe-cluster-policy kube-staging-agent
Expected: list the kube-staging-agent-policy with cluster-admin as target-user

5. zli policy users
Expected: list all existing users in cloud-staging

6. zli policy add-user sai-policy-refactoring-test sebby@commonwealthcrypto.com
Expected: adds Sebby as a user to this policy

7. zli policy delete-user sai-policy-refactoring-test sebby@commonwealthcrypto.com
Expected: deletes Sebby as a user from this policy

8. zli policy groups
Expected: list all existing groups in cloud-staging

9. zli policy add-group sai-policy-refactoring-test Product
Expected: adds Product as a group to this policy

10. zli policy delete-group sai-policy-refactoring-test Product
Expected: deletes Product as a group from this policy

11. zli policy targetusers sai-policy-refactoring-test
Expected: list all existing target users for this policy

12. zli policy add-targetuser sai-policy-refactoring-test testuser
Expected: adds testuser to existing target users for this policy

13. zli policy delete-targetuser sai-policy-refactoring-test testuser
Expected: deletes testuser from existing target users for this policy

14. zli policy targetgroups sai-policy-refactoring-test
Expected: list all existing target groups for this policy

15. zli policy add-targetgroup sai-policy-refactoring-test testgroup
Expected: adds testgroup to existing target groups for this policy

16. zli policy delete-targetgroup sai-policy-refactoring-test testgroup
Expected: deletes testgroup from existing target groups for this policy

**backend branch:** develop
**bzero branch:** develop

#### Ready to run system tests?


## Relevant release note information

Release Notes: 

## Related JIRA tickets

Relates to JIRA: CWC-1542

## Relevant HotFix if applicable

Relevant HotFix PR Number:

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain:
